### ISSUE TITLE
two tokens on jupyterlab notebook start

### ISSUE THREAD

**Issue by vsoch** [jupyter/docker-stacks#701]
This shouldn't make a difference to functionality, but note that the URL given to copy paste to the browser has the "token" variable twice:

![image](https://user-images.githubusercontent.com/814322/44393924-cb01de80-a4ea-11e8-9b09-41ab0327b81e.png)

```bash
[I 09:27:41.849 LabApp] Serving notebooks from local directory: /home/jovyan
[I 09:27:41.849 LabApp] 0 active kernels
[I 09:27:41.849 LabApp] The Jupyter Notebook is running at:
[I 09:27:41.849 LabApp] http://a5933af67d37:8888/?token=88cfd68f3ffada1f1026d5a05246ca65df4751891e090a29
[I 09:27:41.849 LabApp] Use Control-C to stop this server and shut down all kernels (twice to skip confirmation).
[C 09:27:41.850 LabApp] 
    
    Copy/paste this URL into your browser when you connect for the first time,
    to login with a token:
        http://a5933af67d37:8888/?token=88cfd68f3ffada1f1026d5a05246ca65df4751891e090a29&token=88cfd68f3ffada1f1026d5a05246ca65df4751891e090a29
```
They appear to be slightly different - testing didn't work for either for me the first time (not sure why) but when I mapped port 8888 to 8888, I was able to get it working, and it was the second token that did the trick!

```bash
    
    Copy/paste this URL into your browser when you connect for the first time,
    to login with a token:
        http://811a7be7ec57:8888/?token=1ed99d61ff3edb4569b9409fdcdc915b7273d0f45f7febd3&token=1ed99d61ff3edb4569b9409fdcdc915b7273d0f45f7febd3
[I 09:36:50.050 LabApp] 302 GET /?token=88cfd68f3ffada1f1026d5a05246ca65df4751891e090a29 (172.17.0.1) 0.85ms
[I 09:36:50.056 LabApp] 302 GET /lab?token=88cfd68f3ffada1f1026d5a05246ca65df4751891e090a29 (172.17.0.1) 1.03ms
[W 09:37:19.807 LabApp] 401 POST /login?next=%2Flab%3Ftoken%3D88cfd68f3ffada1f1026d5a05246ca65df4751891e090a29 (172.17.0.1) 2.89ms referer=http://0.0.0.0:8888/login?next=%2Flab%3Ftoken%3D88cfd68f3ffada1f1026d5a05246ca65df4751891e090a29
[W 09:37:38.301 LabApp] 401 POST /login?next=%2Flab%3Ftoken%3D88cfd68f3ffada1f1026d5a05246ca65df4751891e090a29 (172.17.0.1) 2.81ms referer=http://0.0.0.0:8888/login?next=%2Flab%3Ftoken%3D88cfd68f3ffada1f1026d5a05246ca65df4751891e090a29
[I 09:37:47.302 LabApp] 302 GET /lab (172.17.0.1) 0.77ms
[I 09:38:17.244 LabApp] 302 GET /login?next=%2Flab&token=1ed99d61ff3edb4569b9409fdcdc915b7273d0f45f7febd3 (172.17.0.1) 1.16ms
[I 09:38:18.958 LabApp] Build is up to date
```
Are the two tokens for different things?

**Comment by parente** [comment]
Interesting. Would you mind sharing the command(s) you used to launch the container?

**Comment by vsoch** [comment]
Sure thing! I have the whole thing here still, thanks to my irresponsibility of properly closing terminals :)

Here is the first run (I think a copy paste from the README.md) that included the pull:

```bash
$ docker run --rm -p 10000:8888 -e JUPYTER_ENABLE_LAB=yes -v "$PWD":/home/jovyan/work jupyter/datascience-notebook:3772fffc4aa4
Unable to find image 'jupyter/datascience-notebook:3772fffc4aa4' locally
3772fffc4aa4: Pulling from jupyter/datascience-notebook
a48c500ed24e: Pulling fs layer 
1e1de00ff7e1: Pulling fs layer 
0330ca45a200: Pulling fs layer 
471db38bcfbf: Pulling fs layer 
0b4aba487617: Pulling fs layer 
d44ea0cd796c: Pulling fs layer 
5ac827d588be: Pull complete 
75a133b7fa5b: Pull complete 
e7175e1009e2: Pull complete 
05f8e0c6e47b: Pull complete 
28dec96a150f: Pull complete 
9b3ade1f5e09: Pull complete 
729fb40bfc4e: Pull complete 
c19915438028: Pull complete 
dbbebc6e5aa4: Pull complete 
fccb4a3fa52c: Pull complete 
d4095ac15122: Pull complete 
be06bbb8d88e: Pull complete 
d96c92f31d53: Pull complete 
d583a4974fef: Pull complete 
040654894752: Pull complete 
a0d70f0a70d8: Pull complete 
eb52839fd778: Pull complete 
75ff689b3ca8: Pull complete 
19814aa1ceb5: Pull complete 
f515e611c48f: Pull complete 
ae06cd746859: Pull complete 
7a8d74ffbe6e: Pull complete 
b6e3af9a6ccf: Pull complete 
Digest: sha256:c2fd52d133dba097e8308afaadbe81a373703869d5c5911f665df4fd115aa72a
Status: Downloaded newer image for jupyter/datascience-notebook:3772fffc4aa4
```
Just pointing this out, since it's a bit hidden, in case you don't want it ignored
```
/usr/local/bin/start-notebook.sh: ignoring /usr/local/bin/start-notebook.d/*
```
And this one too
```
Container must be run with group "root" to update passwd file
```
```
Executing the command: jupyter lab
[I 09:27:41.512 LabApp] Writing notebook server cookie secret to /home/jovyan/.local/share/jupyter/runtime/notebook_cookie_secret
[W 09:27:41.810 LabApp] WARNING: The notebook server is listening on all IP addresses and not using encryption. This is not recommended.
[I 09:27:41.827 LabApp] JupyterLab beta preview extension loaded from /opt/conda/lib/python3.6/site-packages/jupyterlab
[I 09:27:41.827 LabApp] JupyterLab application directory is /opt/conda/share/jupyter/lab
[W 09:27:41.836 LabApp] JupyterLab server extension not enabled, manually loading...
[I 09:27:41.837 LabApp] JupyterLab beta preview extension loaded from /opt/conda/lib/python3.6/site-packages/jupyterlab
[I 09:27:41.837 LabApp] JupyterLab application directory is /opt/conda/share/jupyter/lab
[I 09:27:41.849 LabApp] Serving notebooks from local directory: /home/jovyan
[I 09:27:41.849 LabApp] 0 active kernels
[I 09:27:41.849 LabApp] The Jupyter Notebook is running at:
[I 09:27:41.849 LabApp] http://a5933af67d37:8888/?token=88cfd68f3ffada1f1026d5a05246ca65df4751891e090a29
[I 09:27:41.849 LabApp] Use Control-C to stop this server and shut down all kernels (twice to skip confirmation).
[C 09:27:41.850 LabApp] 
    
    Copy/paste this URL into your browser when you connect for the first time,
    to login with a token:
        http://a5933af67d37:8888/?token=88cfd68f3ffada1f1026d5a05246ca65df4751891e090a29&token=88cfd68f3ffada1f1026d5a05246ca65df4751891e090a29
```
At this point I hadn't even noticed the two tokens because the port mapping didn't seem to be working, but I figured it was user error and just mapped 8888 to 8888 to try again (here comes the control+C)
```
^C[I 09:36:27.752 LabApp] Interrupted...
[I 09:36:27.752 LabApp] Shutting down 0 kernels
```
Now I'm doing it again, same command, different port.
```
$ docker run --rm -p 8888:8888 -e JUPYTER_ENABLE_LAB=yes -v "$PWD":/home/jovyan/work jupyter/datascience-notebook:3772fffc4aa4
/usr/local/bin/start-notebook.sh: ignoring /usr/local/bin/start-notebook.d/*

Container must be run with group "root" to update passwd file
Executing the command: jupyter lab
[I 09:36:39.605 LabApp] Writing notebook server cookie secret to /home/jovyan/.local/share/jupyter/runtime/notebook_cookie_secret
[W 09:36:39.792 LabApp] WARNING: The notebook server is listening on all IP addresses and not using encryption. This is not recommended.
[I 09:36:39.801 LabApp] JupyterLab beta preview extension loaded from /opt/conda/lib/python3.6/site-packages/jupyterlab
[I 09:36:39.801 LabApp] JupyterLab application directory is /opt/conda/share/jupyter/lab
[W 09:36:39.810 LabApp] JupyterLab server extension not enabled, manually loading...
[I 09:36:39.811 LabApp] JupyterLab beta preview extension loaded from /opt/conda/lib/python3.6/site-packages/jupyterlab
[I 09:36:39.811 LabApp] JupyterLab application directory is /opt/conda/share/jupyter/lab
[I 09:36:39.818 LabApp] Serving notebooks from local directory: /home/jovyan
[I 09:36:39.818 LabApp] 0 active kernels
[I 09:36:39.818 LabApp] The Jupyter Notebook is running at:
[I 09:36:39.818 LabApp] http://811a7be7ec57:8888/?token=1ed99d61ff3edb4569b9409fdcdc915b7273d0f45f7febd3
[I 09:36:39.819 LabApp] Use Control-C to stop this server and shut down all kernels (twice to skip confirmation).
[C 09:36:39.819 LabApp] 
    
    Copy/paste this URL into your browser when you connect for the first time,
    to login with a token:
        http://811a7be7ec57:8888/?token=1ed99d61ff3edb4569b9409fdcdc915b7273d0f45f7febd3&token=1ed99d61ff3edb4569b9409fdcdc915b7273d0f45f7febd3
```
This time I did notice it, and actually it's the same token, but twice? You can see the urls I was going to based on the pattern of #fail below. 
```
[I 09:36:50.050 LabApp] 302 GET /?token=88cfd68f3ffada1f1026d5a05246ca65df4751891e090a29 (172.17.0.1) 0.85ms
[I 09:36:50.056 LabApp] 302 GET /lab?token=88cfd68f3ffada1f1026d5a05246ca65df4751891e090a29 (172.17.0.1) 1.03ms
[W 09:37:19.807 LabApp] 401 POST /login?next=%2Flab%3Ftoken%3D88cfd68f3ffada1f1026d5a05246ca65df4751891e090a29 (172.17.0.1) 2.89ms referer=http://0.0.0.0:8888/login?next=%2Flab%3Ftoken%3D88cfd68f3ffada1f1026d5a05246ca65df4751891e090a29
[W 09:37:38.301 LabApp] 401 POST /login?next=%2Flab%3Ftoken%3D88cfd68f3ffada1f1026d5a05246ca65df4751891e090a29 (172.17.0.1) 2.81ms referer=http://0.0.0.0:8888/login?next=%2Flab%3Ftoken%3D88cfd68f3ffada1f1026d5a05246ca65df4751891e090a29
[I 09:37:47.302 LabApp] 302 GET /lab (172.17.0.1) 0.77ms
[I 09:38:17.244 LabApp] 302 GET /login?next=%2Flab&token=1ed99d61ff3edb4569b9409fdcdc915b7273d0f45f7febd3 (172.17.0.1) 1.16ms
[I 09:38:18.958 LabApp] Build is up to date
[I 10:07:08.577 LabApp] Kernel started: 8328933d-8b1d-4163-b08a-56f88117fab8
[I 10:07:09.129 LabApp] Adapting to protocol v5.1 for kernel 8328933d-8b1d-4163-b08a-56f88117fab8
[I 10:07:09.144 LabApp] Adapting to protocol v5.1 for kernel 8328933d-8b1d-4163-b08a-56f88117fab8
```
Again, this doesn't really break anything, but it's probably being appended twice and could lead to some unpredictable bug!

**Comment by iyanmv** [comment]
I can confirm the same issue using `jupyter/datascience-notebook:3772fffc4aa4`. However, with `jupyter/datascience-notebook:latest` it seems to properly give just one token. @vsoch any reason to use that particular tag?

Checking other tags, it looks like image `jupyter/datascience-notebook:621b96ed75cb` was the first one given one token. This image, with respect yours, uses `jupyter 5.6` instead of `jupyter 5.5` so I guess is totally related to this [PR](https://github.com/jupyter/notebook/pull/3656/). See also the [changelog](https://jupyter-notebook.readthedocs.io/en/stable/changelog.html) (Fixing problems section):

> Don’t duplicate token in displayed URL (PR #3656)

So, basically not a docker-stacks issue, but an old jupyter bug that was fixed in version 5.6. Use a more recent tag or just ignore the duplicated token if you need that particular tag :wink:

**Comment by vsoch** [comment]
Ah excellent! That tag is the one in the README.md, so it's the one I copy pasted (and honestly didn't even realize it was there with the long command!). Maybe change the tag in the README.md to be latest (or a more recent one without the bug) as a suggestion to resolve the issue?

I can also confirm all is well using the latest tag!

![image](https://user-images.githubusercontent.com/814322/44514034-fe7c6e80-a68c-11e8-8941-04a4e07abdef.png)

Also as a suggestion, it might make sense (for the example to have the equivalent port mapped, e.g `8888:8888` because half the users (myself included, even noticing in advance) are going to see and copy paste the url, and then be utterly confused when it doesn't open (because the message tells me 8888, but I needed to open the mapped port 10000). On the other hand, when it works the first time and you see that Jupyter logo spinning, and then the panel? :sparkles: magic :sparkles:

**Comment by iyanmv** [comment]
I think one of the purposes of the second example in the [README.md](https://github.com/jupyter/docker-stacks/blob/8ccdfc1da8d5048874bbc202aefbf16136aaba60/README.md) is to show that it is easy to use a different host port (first example does use `-p 8888:8888`). Maybe it is confused if read alone, but it totally makes sense in the context of the previous example.

I agree that changing the tag in the example is a good idea. Maybe a good option is to add an additional commit to PR #693 since other documentation files have been changed.

**Comment by parente** [comment]
This issue has inactive for some time now and so I'm going to close it. If someone would like to send a PR to make the documentation clearer, I'll be happy to review it.

**Comment by vsoch** [comment]
holy cow, like 7 months and nobody changed a version string in the README? Holy cow man, I'll do it. The version clearly has a bug and needs to be updated. @parente please reopen and assign me to this issue.

**Comment by vsoch** [comment]
Nevermind, we're good! I clicked a link to an old readme. Relief! I couldn't believe it.
